Add a concrete `util::Error` type (step 3 of refactoring)

jtgeibel opened the pull request:
This PR may be easier to review commit by commit.  Overall this series continues the previous refactoring work in #1920 and #1929.  In particular, a concrete `util::Error` error type is introduced.  The code is updated to follow a style where:

* The new concrete type `util::Error` is great for code that is not part of the request/response lifecycle.  Binaries have been converted to `fn main() -> Result<(), Error>` where applicable.  This avoids pulling in the unnecessary infrastructure to convert errors into a user facing JSON responses (relative to `AppError`).
  * There are now 13 results across 6 files for `, Error>`.  Usage occurs in `src/bin/`, `src/boot/`, and `src/uploaders.rs`.
* `diesel::QueryResult` - There is a lot of code that only deals with query errors.  If only one type of error is possible in a function, using that specific error is preferable to the more general `util::Error`.  This is especially common in model code.
  * There are now 49 results across 16 files for `QueryResult<`.  Most usage is in `src/models/`, with some usage in `src/publish_rate_limit.rs`, `src/tasks/`, and `src/tests/`.
* `util::errors::AppResult` - Some failures should be converted into user facing JSON responses.  This error type is more dynamic and is box allocated.  Low-level errors are typically not converted to user facing errors and most usage is within the models, controllers, and middleware layers (85 of 102 results below).
  * There are now 102 results across 36 files for `AppResult<`.  Most usage is in `src/controllers/` and `src/models/`, with additional usage in `src/middleware/`, `src/{email,github,publish_rate_limit,router,uploaders}.rs`, and `src/{tests,util}/`.

Finally, a `middleware::Result` type is introduced for use in middleware and router code.

The behavior of `dyn AppError` and `chain_error` is a remaining area for improvement.  There may be cases where we want to wrap an `internal` error in a user facing error (like `cargo_err` or `server_err`) and still have the internal error make it up to the logging middleware.  However, I don't see an easy way to propagate this up the middleware stack (without encoding it in an internal response header that we later remove, :sweat:).

r? @smarnach

smarnach commented:
I plan to review this on 27 December, but this plan is subject to change…

bors commented:
:umbrella: The latest upstream changes (presumably #1931) made this pull request unmergeable. Please resolve the merge conflicts.

smarnach commented:
I'm sorry I did not get to this today, and I won't have time until 2020-01-05.

bors commented:
:umbrella: The latest upstream changes (presumably #2072) made this pull request unmergeable. Please resolve the merge conflicts.

jtgeibel commented:
> I'm sorry I did not get to this today, and I won't have time until 2020-01-05.

No problem at all.  When you have a chance, would you start with PR #2071.  It's a small fix and I'd like to get it deployed, just want a 2nd set of eyes on it.  Thanks!

bors commented:
:umbrella: The latest upstream changes (presumably #2066) made this pull request unmergeable. Please resolve the merge conflicts.

smarnach commented on src/bin/enqueue-job.rs:
```diff
@@ -1,29 +1,20 @@
+            Ok(tasks::dump_db(database_url, target_name).enqueue(&conn)?)
         }
-        other => Err(cargo_err(&format!("Unrecognized job type `{}`", other))),
+        other => Err(Error::from(format!("Unrecognized job type `{}`", other))),
```
You could use `?` instead of `Error::from` here as well, as you did in the other match arms. (I don't have a strong opinion either way.)

I'm glad you could get rid of the helper trait here. :)

jtgeibel replied:
It looks like clippy doesn't like `Err(format!("Unrecognized job type `{}`", other))?` (the `clippy::try_err` lint), and the suggestion is to use `.into()`.  I don't have a strong opinion between `Error::from` and `into` here, so I'll leave it as-is.

jtgeibel commented on src/background_jobs.rs:
```diff
@@ -1,21 +1,22 @@
-use swirl::PerformError;
+
+use diesel::r2d2::PoolError;
+use swirl::errors::PerformError;
```
It looks like I removed it at one point in the series, and then added it back in, so I probably copied the longer path from an error message.

I added a commit converting 3 places to use the short path (now 6) for consistency.

jtgeibel commented:
> It would be great if you could introduce some high-level documentation in a doc comment at the beginning of error.rs, explaining the use cases for the main error or result types – basically a summary of what you put in the PR description.

I had been holding off on adding high-level documentation until things settled down, but I think this is now a good time to add these recommendations as documentation.  (Especially since "I'll add some docs in a future PR" always takes longer than expected to get back around to.)  I've rebased the PR, and added a commit adapting part of the PR description into module docs.

Thanks for the review and feedback!

@bors r+

bors commented:
:pushpin: Commit 541f710520043023e648ed07a3358e285ebead34 has been approved by `jtgeibel`

<!-- @bors r=jtgeibel 541f710520043023e648ed07a3358e285ebead34 -->
<!-- homu: {"type":"Approved","sha":"541f710520043023e648ed07a3358e285ebead34","approver":"jtgeibel"} -->

bors commented:
:hourglass: Testing commit 541f710520043023e648ed07a3358e285ebead34 with merge a251368963d55ff12ecb65249cfe96950deb867b...
<!-- homu: {"type":"BuildStarted","head_sha":"541f710520043023e648ed07a3358e285ebead34","merge_sha":"a251368963d55ff12ecb65249cfe96950deb867b"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/crates.io/builds/143186626)
Approved by: jtgeibel
Pushing a251368963d55ff12ecb65249cfe96950deb867b to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jtgeibel","base_ref":"master","builders":{"checks-travis":"https://travis-ci.com/rust-lang/crates.io/builds/143186626"},"merge_sha":"a251368963d55ff12ecb65249cfe96950deb867b"} -->
